src/Microsoft.ML.Transforms/Text/TextFeaturizingEstimator.cs

```diff
@@ -115,67 +115,28 @@ internal sealed class Arguments : TransformInputBase
+            internal INgramExtractorFactoryFactory CharFeatureExtractorFactory = new NgramExtractorTransform.NgramExtractorArguments() { NgramLength = 3, AllLengths = false };
+
+            public WordBagEstimator.Options CharFeatureExtractor;
 
             [Argument(ArgumentType.AtMostOnce, HelpText = "Normalize vectors (rows) individually by rescaling them to unit norm.", ShortName = "norm", SortOrder = 13)]
             public NormFunction VectorNormalizer = NormFunction.L2;
-        }
 
-        /// <summary>
-        /// Advanced options for the <see cref="TextFeaturizingEstimator"/>.
-        /// </summary>
-        public sealed class Options
-        {
-#pragma warning disable MSML_NoInstanceInitializers // No initializers on instance fields or properties
-            /// <summary>
-            /// Dataset language.
-            /// </summary>
-            public Language TextLanguage { get; set; } = DefaultLanguage;
-            /// <summary>
-            /// Casing used for the text.
-            /// </summary>
-            public CaseNormalizationMode TextCase { get; set; } = CaseNormalizationMode.Lower;
-            /// <summary>
-            /// Whether to keep diacritical marks or remove them.
-            /// </summary>
-            public bool KeepDiacritics { get; set; } = false;
-            /// <summary>
-            /// Whether to keep punctuation marks or remove them.
-            /// </summary>
-            public bool KeepPunctuations { get; set; } = true;
-            /// <summary>
-            /// Whether to keep numbers or remove them.
-            /// </summary>
-            public bool KeepNumbers { get; set; } = true;
-            /// <summary>
-            /// Whether to output the transformed text tokens as an additional column.
-            /// </summary>
-            public bool OutputTokens { get; set; } = false;
-            /// <summary>
-            /// Vector Normalizer to use.
-            /// </summary>
-            public NormFunction VectorNormalizer { get; set; } = NormFunction.L2;
-            /// <summary>
-            /// Whether to use stop remover or not.
-            /// </summary>
-            public bool UseStopRemover { get; set; } = false;
-            /// <summary>
-            /// Whether to use char extractor or not.
-            /// </summary>
-            public bool UseCharExtractor { get; set; } = true;
-            /// <summary>
-            /// Whether to use word extractor or not.
-            /// </summary>
-            public bool UseWordExtractor { get; set; } = true;
-#pragma warning restore MSML_NoInstanceInitializers // No initializers on instance fields or properties
+            public Options()
+            {
+                WordFeatureExtractor = new WordBagEstimator.Options();
+                CharFeatureExtractor = new WordBagEstimator.Options() { NgramLength = 3, AllLengths = false };
```

Comment:
Shouldn't these fields initialized to null by default? And then fallback to WordFeatureExtractorFactory / CharFeatureExtractorFactory whenever you try to use them and they are not defined. This way logic from entrypoint / maml will be similar to API.

Reply:
`WordFeatureExtractor` and `CharFeatureExtractor` both can be null. When they are null its means do not apply the respective transforms. So, there is no way to detect if null was set by user or its a default value. 

Right now, WordFeatureExtractorFactory / CharFeatureExtractorFactory is preferred if the maml is used.

---
In reply to: [264465314](https://github.com/dotnet/machinelearning/pull/2911#discussion_r264465314) [](ancestors = 264465314)